cpp/test/packed/stream_writer_test.cpp

```diff
@@ -0,0 +1,139 @@
+      columns.push_back(int64_array);
+      fields.push_back(arrow::field("int64_" + std::to_string(i), arrow::int64()));
+    }
+
+    arrow::StringBuilder str_builder;
+    ASSERT_TRUE(
+        str_builder.AppendValues({std::string(10000, 'a'), std::string(10000, 'b'), std::string(10000, 'c')}).ok());
+    std::shared_ptr<arrow::Array> str_array;
+    ASSERT_TRUE(str_builder.Finish(&str_array).ok());
+    fields.push_back(arrow::field("large_str", arrow::utf8()));
+    columns.push_back(str_array);
+    schema_ = arrow::schema(fields);
+
+    record_batch_ = arrow::RecordBatch::Make(schema_, 3, columns);
+
+    fs_ = std::unique_ptr<arrow::fs::FileSystem>(new arrow::fs::LocalFileSystem());
+    // ASSERT_AND_ASSIGN(fs_, BuildFileSystem("file:///tmp/", &path));
+    props_ = *parquet::default_writer_properties();
+    memory_limit_ = 1024 * 1024;  // 1MB for testing
+  }
+
+  void TearDown() override { std::filesystem::remove_all(file_path_); }
+
+  std::shared_ptr<arrow::Schema> schema_;
+  std::shared_ptr<arrow::RecordBatch> record_batch_;
+  std::unique_ptr<arrow::fs::FileSystem> fs_;
+  std::string file_path_;
+  parquet::WriterProperties props_;
+  size_t memory_limit_;
+};
+
+TEST_F(StreamWriterTest, Write) {
+  StreamWriter writer(memory_limit_, schema_, *fs_, file_path_, props_);
+  EXPECT_TRUE(writer.Init(record_batch_).ok());
+  EXPECT_TRUE(writer.Write(record_batch_).ok());
+  EXPECT_TRUE(writer.Close().ok());
+
+  std::set<int> needed_columns = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+  std::vector<ColumnOffset> column_offsets = {
+      ColumnOffset(0, 0),
+  };
+  for (int i = 0; i < 10; ++i) {
+    column_offsets.emplace_back(1, i);
+  }
+
+  auto paths = std::vector<std::string>{file_path_ + "/0", file_path_ + "/1"};
+
+  // after writing, the column of large_str is in 0th file, and the last int64 10 columns are in 1st file
+  std::vector<std::shared_ptr<arrow::Field>> fields = {arrow::field("large_str", arrow::utf8())};
+  for (int i = 1; i < 11; ++i) {
+    fields.push_back(arrow::field("int64_" + std::to_string(i), arrow::int64()));
+  }
+  auto new_schema = arrow::schema(fields);
+
+  PackedRecordBatchReader pr(*fs_, paths, new_schema, column_offsets, needed_columns);
+
+  EXPECT_EQ(11, pr.schema()->num_fields());
+
+  ASSERT_AND_ARROW_ASSIGN(auto table, pr.ToTable());
+  ASSERT_AND_ARROW_ASSIGN(auto combined_table, table->CombineChunks());
+
+  auto str_res = std::dynamic_pointer_cast<arrow::StringArray>(combined_table->GetColumnByName("large_str")->chunk(0));
+
+  std::vector<std::string> strs;
+  strs.reserve(str_res->length());
+  for (int i = 0; i < str_res->length(); ++i) {
+    strs.push_back(str_res->GetString(i));
+  }
+
+  EXPECT_THAT(strs, testing::ElementsAre(std::string(10000, 'a'), std::string(10000, 'b'), std::string(10000, 'c'),
+                                         std::string(10000, 'a'), std::string(10000, 'b'), std::string(10000, 'c')));
+  for (int i = 1; i < 11; ++i) {
```

Comment:
It may be necessary to add a test that read and write files that larger than the buffer size, to verify the state transitions on advancing new buffers.

Reply:
updated the test